File: emisor/link/crc32.py
# Capa de Enlace - CRC-32 | Emisor
# Responsabilidad: calcular el CRC-32 del mensaje y concatenarlo al frame.

# Proceso (división binaria por XOR):
#   1. Agregar 32 ceros al final del mensaje
#   2. Alinear el polinomio al bit más significativo del mensaje
#   3. Realizar XOR y avanzar al siguiente '1' hasta que queden solo 32 bits
#   4. El remainder de 32 bits es el CRC
#   5. Concatenar el CRC al mensaje original


# Polinomio estándar CRC-32
POLINOMIO_CRC32 = "100000100110000010001110110110111"

# Recibe el mensaje en bits y retorna el frame completo: mensaje_original + 32 bits de CRC.
def calcular_crc32(bits_mensaje: str) -> str:
    # Validar que el mensaje tenga al menos 32 bits; si no, agregar padding
    if len(bits_mensaje) < 32:
        padding = 32 - len(bits_mensaje)
        bits_mensaje = ('0' * padding) + bits_mensaje
        print(f"[ENLACE-CRC32] Mensaje menor a 32 bits, se agregaron {padding} bits de padding")

    # Paso 1: agregar 32 ceros al final
    dividendo = bits_mensaje + ('0' * 32)

    print(f"[ENLACE-CRC32] Mensaje ({len(bits_mensaje)} bits) + 32 ceros = {len(dividendo)} bits")

    # Paso 2 y 3: división binaria por XOR
    remainder = _division_binaria_xor(dividendo, POLINOMIO_CRC32)

    remainder = remainder.zfill(32)

    print(f"[ENLACE-CRC32] CRC-32 calculado: {remainder}")

    # Paso 5: frame = mensaje original + CRC
    frame = bits_mensaje + remainder
    print(f"[ENLACE-CRC32] Frame final ({len(frame)} bits): {frame}")
    return frame

# Realiza la división binaria por XOR y retorna el remainder final.
def _division_binaria_xor(dividendo: str, divisor: str) -> str:
    len_divisor = len(divisor)
    resultado = list(dividendo)

    for i in range(len(dividendo) - len_divisor + 1):
        # Solo se opera si el bit actual es '1'
        if resultado[i] == '1':
            for j in range(len_divisor):
                # XOR bit a bit
                resultado[i + j] = '0' if resultado[i + j] == divisor[j] else '1'

    remainder = "".join(resultado[-(len_divisor - 1):])
    return remainder

File: receptor/link/crc32.go
// Capa de Enlace - CRC-32 | Receptor
// Responsabilidad: verificar la integridad del frame recibido mediante CRC-32.
//
// Proceso de verificación
//  1. Tomar el frame completo (mensaje original + 32 bits CRC)
//  2. Dividirlo entre el polinomio estándar usando XOR
//  3. Si el remainder es 0 → no hay errores
//     Si el remainder != 0 → se detectó un error
//  4. CRC-32 NO puede corregir errores, solo detectarlos
//  5. Si no hay error, extraer el mensaje original (frame sin los últimos 32 bits)

package link

import (
	"fmt"
	"strings"
)

// Polinomio estándar CRC-32
const polinomioCRC32 = "100000100110000010001110110110111"

// VerificarCRC32 recibe el frame completo (mensaje + 32 bits CRC).
// Retorna los bits del mensaje original (sin el CRC) y un error si se detectaron errores.
func VerificarCRC32(frame string) (string, error) {
	n := len(frame)
	fmt.Printf("[ENLACE-CRC32] Frame recibido (%d bits)\n", n)

	if n < 32 {
		return "", fmt.Errorf("frame demasiado corto (%d bits), se esperan al menos 32 bits de CRC", n)
	}

	// Verificar: dividir el frame entre el polinomio
	remainder := divisionBinariaXOR(frame, polinomioCRC32)
	// El remainder debe ser 32 bits de ceros si no hay error
	remainder = fmt.Sprintf("%032s", remainder) // padding a 32 bits

	fmt.Printf("[ENLACE-CRC32] Remainder calculado: %s\n", remainder)

	// Verificar si el remainder es todo ceros
	solosCeros := true
	for _, c := range remainder {
		if c != '0' {
			solosCeros = false
			break
		}
	}

	if !solosCeros {
		return "", fmt.Errorf("error detectado (remainder != 0). No es posible corregir con CRC-32")
	}

	fmt.Println("[ENLACE-CRC32] No se detectaron errores.")

	// Extraer solo el mensaje original (sin los últimos 32 bits de CRC)
	mensajeBits := frame[:n-32]
	fmt.Printf("[ENLACE-CRC32] Bits de mensaje extraídos (%d bits)\n", len(mensajeBits))
	return mensajeBits, nil
}

// divisionBinariaXOR realiza la división binaria por XOR.
// Retorna el remainder de la división.
func divisionBinariaXOR(dividendo, divisor string) string {
	lenDivisor := len(divisor)
	resultado := []byte(dividendo)

	for i := 0; i <= len(resultado)-lenDivisor; i++ {
		if resultado[i] == '1' {
			for j := 0; j < lenDivisor; j++ {
				if resultado[i+j] == divisor[j] {
					resultado[i+j] = '0'
				} else {
					resultado[i+j] = '1'
				}
			}
		}
	}

	remainder := string(resultado[len(resultado)-(lenDivisor-1):])
	return strings.TrimLeft(remainder, "0")
}
